aemcmc/negative_binomial_horseshoe_gibbs.py

```diff
@@ -0,0 +1,190 @@
+import aesara
+import aesara.tensor as at
+from aesara.ifelse import ifelse
+from aesara.tensor.var import TensorVariable
+
+from aemcmc.dists import (
+    multivariate_normal_cong2017,
+    multivariate_normal_rue2005,
+    polyagamma,
+)
+
+
+def _update_beta_low_dimension(rng, omega, lambda2tau2_inv, X, z):
+    Q = X.T @ (omega[:, None] * X)
+    indices = at.arange(Q.shape[1])
+    Q = at.subtensor.set_subtensor(
+        Q[indices, indices],
+        at.diag(Q) + lambda2tau2_inv,
+    )
+    return multivariate_normal_rue2005(rng, X.T @ (omega * z), Q)
+
+
+def _update_beta_high_dimension(rng, omega, lambda2tau2_inv, X, z):
+    return multivariate_normal_cong2017(rng, lambda2tau2_inv, omega, X, z)
+
+
+def _update_beta(rng, omega, lambda2tau2_inv, X, z):
+    return ifelse(
+        X.shape[1] > X.shape[0],
+        _update_beta_high_dimension(rng, omega, lambda2tau2_inv, X, z),
+        _update_beta_low_dimension(rng, omega, lambda2tau2_inv, X, z),
+    )
+
+
+# NOTE: invGamma(1, a) === 1 / Exp(a), where a is the scale/mean of the exponential
+def horseshoe_step(srng, beta, sigma2, lambda2_inv, tau2_inv):
+    upsilon_inv = srng.exponential(1 + lambda2_inv)
+    zeta_inv = srng.exponential(1 + tau2_inv)
+
+    beta2 = beta * beta
+    lambda2_inv_new = srng.exponential(upsilon_inv + 0.5 * beta2 * tau2_inv / sigma2)
+    tau2_inv_new = srng.gamma(
+        0.5 * (beta.shape[0] + 1),
+        zeta_inv + 0.5 * (beta2 * lambda2_inv_new).sum() / sigma2,
+    )
+    return lambda2_inv_new, tau2_inv_new, upsilon_inv, zeta_inv
+
+
+def horseshoe_nbinom_gibbs(
+    srng: at.random.RandomStream,
+    X: TensorVariable,
+    y: TensorVariable,
+    h: TensorVariable,
+    beta_init: TensorVariable,
+    local_shrinkage_init: TensorVariable,
+    global_shrinkage_init: TensorVariable,
+    n_samples: TensorVariable,
+):
+    r"""
+    Compose a symbolic graph that describes the gibbs sampler of the negative
+    binomial regression with a HorseShoe prior on the regression coefficients.
+
+    The implemenation follows the sampler desribed in [1].
+
+    Parameters
+    ----------
+    srng: at.random.RandomStream
```

Comment:
The type hints are enough; we don't need to maintain type information in the docstring as well.  If anything, we should automatically generate the type information from the type hints at some point; otherwise, we're doing double the work.